File: services/exam-service/src/main/java/io/github/bardiakz/exam_service/entity/Submission.java
package io.github.bardiakz.exam_service.entity;

import jakarta.persistence.*;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "submissions", uniqueConstraints = {
        @UniqueConstraint(columnNames = {"exam_id", "student_id"})
})
public class Submission {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private Long examId;

    @Column(nullable = false)
    private String studentId;

    @OneToMany(mappedBy = "submission", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Answer> answers = new ArrayList<>();

    @Column(nullable = false)
    private LocalDateTime submittedAt;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private SubmissionStatus status = SubmissionStatus.IN_PROGRESS;

    @Column
    private Integer totalScore;

    @Column
    private Integer obtainedScore;

    @Column
    private LocalDateTime gradedAt;

    @Column
    private String feedback;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt = LocalDateTime.now();

    public enum SubmissionStatus {
        IN_PROGRESS,
        SUBMITTED,
        GRADED
    }
}

File: frontend/um_flutter/lib/models/submission.dart
class Submission {
  final int id;
  final int examId;
  final String studentId;
  final String status;
  final int? totalScore;
  final int? obtainedScore;
  final DateTime? submittedAt;
  final DateTime? gradedAt;
  final String? feedback;

  Submission({
    required this.id,
    required this.examId,
    required this.studentId,
    required this.status,
    this.totalScore,
    this.obtainedScore,
    this.submittedAt,
    this.gradedAt,
    this.feedback,
  });

  factory Submission.fromJson(Map<String, dynamic> json) {
    return Submission(
      id: json['id'],
      examId: json['examId'],
      studentId: json['studentId'],
      status: json['status'],
      totalScore: json['totalScore'],
      obtainedScore: json['obtainedScore'],
      submittedAt: json['submittedAt'] != null
          ? DateTime.parse(json['submittedAt'])
          : null,
      gradedAt:
          json['gradedAt'] != null ? DateTime.parse(json['gradedAt']) : null,
      feedback: json['feedback'],
    );
  }
}
